File: src/main/java/com/leet/medium/Permutations.exs
defmodule Solution do
  def solve(nums, size, cur \\ [], done \\ %{}, done_count \\ 0, ans \\ []) do
    if done_count == size do
      [cur | ans]
    else
      Enum.reduce(nums, ans, fn num, acc ->
        if !Map.get(done, num, false) do
          solve(nums, size, [num | cur], Map.put(done, num, true), done_count + 1, acc)
        else
          acc
        end
      end)
    end
  end

  @spec permute(nums :: [integer]) :: [[integer]]
  def permute(nums) do
    solve(nums, length(nums))
  end
end

IO.inspect(Solution.permute([1, 2, 3, 4, 5, 6]))

File: src/main/java/com/leet/medium/Permutations.go
package main

type Permute struct {
	last int
	nums []int
	ans  [][]int
}

func (this *Permute) solve(ans []int, done []bool, pos int, state int) {
	if state == this.last {
		clone := make([]int, len(this.nums))
		copy(clone, ans)
		this.ans = append(this.ans, clone)
		return
	}

	for idx, num := range this.nums {
		if !done[idx] {
			done[idx] = true
			ans[pos] = num
			this.solve(append(ans, num), done, pos+1, state|(1<<idx))
			done[idx] = false
		}
	}
}

func permute(nums []int) [][]int {
	p := Permute{
		last: (1 << len(nums)) - 1,
		nums: nums,
		ans:  [][]int{},
	}
	p.solve(make([]int, len(nums)), make([]bool, len(nums)), 0, 0)
	return p.ans
}
